python/dgl/transform/module.py

```diff
@@ -0,0 +1,790 @@
+##
+#   Copyright 2019-2021 Contributors
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+"""Modules for transform"""
+
+from .. import convert
+from .. import backend as F
+from . import functional
+
+__all__ = [
+    'BaseTransform',
+    'AddSelfLoop',
+    'RemoveSelfLoop',
+    'AddReverse',
+    'ToSimple',
+    'LineGraph',
+    'KHopGraph',
+    'AddMetaPaths',
+    'KNNGraph',
+    'Compose'
+]
+
+def update_graph_structure(g, data_dict, copy_edata=True):
+    r"""
+
+    Description
+    -----------
+    Update the structure of a graph.
+
+    Parameters
+    ----------
+    g : DGLGraph
+        The graph to update.
+    data_dict : graph data
+        The dictionary data for constructing a heterogeneous graph.
+    copy_edata : bool
+        If True, it will copy the edge features to the updated graph.
+
+    Returns
+    -------
+    DGLGraph
+        The updated graph.
+    """
+    device = g.device
+    idtype = g.idtype
+    num_nodes_dict = dict()
+
+    for ntype in g.ntypes:
+        num_nodes_dict[ntype] = g.num_nodes(ntype)
+
+    new_g = convert.heterograph(data_dict, num_nodes_dict=num_nodes_dict,
+                                idtype=idtype, device=device)
+
+    # Copy features
+    for ntype in g.ntypes:
+        for key, feat in g.nodes[ntype].data.items():
+            new_g.nodes[ntype].data[key] = feat
+
+    if copy_edata:
+        for c_etype in g.canonical_etypes:
+            for key, feat in g.edges[c_etype].data.items():
+                new_g.edges[c_etype].data[key] = feat
+
+    return new_g
+
+class BaseTransform:
+    r"""
+
+    Description
+    -----------
+    An abstract class for writing transforms.
+    """
+    def __call__(self, g):
+        raise NotImplementedError
+
+    def __repr__(self):
+        return self.__class__.__name__ + '()'
+
+class AddSelfLoop(BaseTransform):
+    r"""
+
+    Description
+    -----------
+    Add self-loops for each node in the graph and return a new graph.
+
+    For heterogeneous graphs, self-loops are added only for edge types with same
+    source and destination node types.
+
+    Parameters
+    ----------
+    remove_first : bool, optional
+        If True, it will first remove self-loops to prevent duplicate self-loops.
+    add_self_etypes : bool, optional
```

Comment:
`new_etype=False`

Reply:
Then will this still add reverse canonical edge types for asymmetric canonical edge types?

Reply:
It will.

Reply:
You could change it to a more explicit name.

Reply:
Changed.